### PR TITLE
[RFC] use statistic to order prewhere conditions better

### PR DESCRIPTION
**Pull request by hanfei1991**
### What is statistic

an explaination from SQL Server https://learn.microsoft.com/en-us/sql/relational-databases/statistics/statistics?view=sql-server-ver16

We use `tdigest` as a pratical histogram

### How to create / manipulate statistic in ClickHouse

we treat statistic as a property of a table, such as `CODEC`, `TTL` ...

but we store and manipulate statistic seperately, such as `INDEX`, `PROJECTION`

**create**

```[sql]
CREATE TABLE test (
    a Int64 STATISTIC type (tdigest),
    b Int64 STATISTIC type (tdigest)
) ...
```

**manipulate**

```[sql]
ALTER TABLE test ADD / DROP / CLEAR / MATERIALIZE STATISTIC a, b TYPE tdigest
```

### How to store it in a part

we store a single file containing all types of statistics for every column which has statistic.

image

and
- we write statistic automatically when inserting / merging to a new part

### how do we use statistic in where optimizer

- only work for integers
- when `prewhere condition` is like `a < 5 and b > 1 and c < 4.0`, then we try to re-order them accroding to the selectivity
- read statistic from file systems every time

### more works need to do for this PR:

- [x] doc + rfc
- [x] tests
- [x] validate for statistic create
- [x] collect statistic by partition pruner
- [x] more comments
- [ ] serialize statistic files as above described (will be implemented in the next PR)
- [x] manipulate statistic when column changed
  - [x] drop/clear statistic when column is dropped/cleared
  - [x] rename ...
  - [x] do not modify column when the changed type conflicts with statistic
- [x] implement `lessThan` for `TDigest`

hope that I could control the + lines within 2000 :smiley_cat: 

### what to do in the future

https://github.com/ClickHouse/ClickHouse/issues/55065

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
use statistic to order prewhere conditions better

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit 2b1f2f1de2625f83c944f5a26e8ad5f0a3a50b09 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53240/2b1f2f1de2625f83c944f5a26e8ad5f0a3a50b09/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

File: src/Storages/Statistic/Estimator.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        auto function_argument = function_node.getArgumentAt(i);
+        auto subresult = extractSingleColumn(function_argument);
+        if (subresult == std::nullopt)
+            continue;
+        else if (subresult == "")
+            return "";
+        else if (result == std::nullopt)
+            result = subresult;
+        else if (result.value() != subresult.value())
+            return "";
```
Comment (CurtizJ) [inline src/Storages/Statistic/Estimator.cpp:34]: It looks a bit messy. Can we leave only returning of `std::nullopt` in if there is no suitable column?

File: src/Storages/Statistic/Statistic.h [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    virtual UInt64 count() = 0;
+
+protected:
+
+    StatisticDescription stat;
+
+};
+
+/// TDigestStatistic is a kind of histogram.
+class TDigestStatistic : public IStatistic
```
Comment (CurtizJ) [inline src/Storages/Statistic/Statistic.h:63]: Consider to move it to separate file.

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 3]
```diff
@@ -3125,6 +3181,17 @@ void MergeTreeData::checkAlterIsPossible(const AlterCommands & commands, Context
...
+                    const auto & old_column = old_metadata.getColumns().get(command.column_name);
+                    if (old_column.stat)
+                    {
+                        const auto & new_column = new_metadata.getColumns().get(command.column_name);
+                        if (!old_column.type->equals(*new_column.type))
+                            throw Exception(ErrorCodes::ALTER_OF_COLUMN_IS_FORBIDDEN,
+                                            "ALTER types of column {} with statistic is not not safe "
+                                            "because it can change the representation of statistic",
+                                            backQuoteIfNeed(command.column_name));
+                    }
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply src/Storages/MergeTree/MergeTreeData.cpp:3194]: yes, I should add it to the backlog

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 4]
```diff
@@ -444,6 +445,55 @@ StoragePolicyPtr MergeTreeData::getStoragePolicy() const
...
+    if (parts.empty())
+    {
+        return {};
+    }
+
+    ASTPtr expression_ast;
+    Block virtual_columns_block = getBlockWithVirtualPartColumns(parts, true /* one_part */);
+    //
+    // Generate valid expressions for filtering
+    bool valid = VirtualColumnUtils::prepareFilterBlockWithQuery(query_info.query, local_context, virtual_columns_block, expression_ast);
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply src/Storages/MergeTree/MergeTreeData.cpp]: removed it

File: src/Storages/ColumnsDescription.h [hunk 5]
```diff
@@ -83,6 +84,7 @@ struct ColumnDescription
     String comment;
     ASTPtr codec;
     ASTPtr ttl;
+    std::optional<StatisticDescription> stat;
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply src/Storages/ColumnsDescription.h:87]: done

File: src/Storages/AlterCommands.cpp [hunk 6]
```diff
@@ -555,6 +589,42 @@ void AlterCommand::apply(StorageInMemoryMetadata & metadata, ContextPtr context)
...
     }
+    else if (type == ADD_STATISTIC)
+    {
+        for (const auto & statistic_column_name : statistic_columns)
+        {
+            if (!metadata.columns.has(statistic_column_name))
+            {
+                throw Exception(ErrorCodes::ILLEGAL_STATISTIC, "Cannot add statistic {} with type {}: this column is not found", statistic_column_name, statistic_type);
+            }
+            if (metadata.columns.get(statistic_column_name).stat)
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply src/Storages/AlterCommands.cpp]: yes type should be check as well

File: src/Storages/AlterCommands.cpp [hunk 7]
```diff
@@ -293,6 +310,23 @@ std::optional<AlterCommand> AlterCommand::parse(const ASTAlterCommand * command_
...
+        command.statistic_columns = ast_stat_decl.getColumnNames();
+        command.statistic_type = ast_stat_decl.type;
+        command.if_exists = command_ast->if_exists;
+        command.clear = command_ast->clear_statistic;
+
+        if (command_ast->partition)
+            command.partition = command_ast->partition;
+
+        return command;
+    }
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply src/Storages/AlterCommands.cpp:329]: no, the `materialize_(index|projection)` is not here neither. this function is only for stmt which changes the meta

File: src/Parsers/ASTStatisticDeclaration.h [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+class ASTFunction;
+
+/** name BY columns TYPE typename(args) in create query
+  */
+class ASTStatisticDeclaration : public IAST
+{
+public:
+    IAST * columns;
+    /// TODO type should be a list of ASTFunction, for example, 'tdigest(256), hyperloglog(128)', etc.
+    String type;
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply src/Parsers/ASTStatisticDeclaration.h:17]: will do it when we support second stat type

File: docs/en/engines/table-engines/mergetree-family/mergetree.md [hunk 9]
```diff
@@ -1355,3 +1355,33 @@ In this sample configuration:
...
+    b Float64
+)
+ENGINE = MergeTree
+ORDER BY a
+```
+
+We can also manipulate statistics with `ATLER` statements.
+
+```sql
+ATLER TABLE example_table ADD STATISTIC b TYPE tdigest;
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply docs/en/engines/table-engines/mergetree-family/mergetree.md]: done

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 10]
```diff
@@ -444,6 +445,55 @@ StoragePolicyPtr MergeTreeData::getStoragePolicy() const
     return storage_policy;
 }
 
+ConditionEstimator MergeTreeData::getConditionEstimatorByPredicate(const SelectQueryInfo & query_info, ContextPtr local_context) const
+{
+    if (!local_context->getSettings().allow_statistic_optimize)
+        return {};
+
+    auto parts = getDataPartsVectorForInternalUsage();
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply src/Storages/MergeTree/MergeTreeData.cpp]: done

File: src/Storages/MergeTree/IMergeTreeDataPart.cpp [hunk 11]
```diff
@@ -633,6 +634,31 @@ String IMergeTreeDataPart::getColumnNameWithMinimumCompressedSize(bool with_subc
...
+        if (!metadata_manager->exists(file_name))
+        {
+            LOG_INFO(storage.log, "Cannot find stats file {}", file_path);
+            continue;
+        }
+        auto stat_file = metadata_manager->read(file_name);
+        CompressedReadBuffer compressed_buffer(*stat_file);
+        stat->deserialize(compressed_buffer);
+        result.push_back(stat);
+    }
```
[the remark this answers is not in the document]
Comment (hanfei1991) [reply src/Storages/MergeTree/IMergeTreeDataPart.cpp:658]: one column has one statistic file containing all kinds of statistics.

**Comment by JackyWoo** [comment]
hi @hanfei1991 I am intresting with your PR. I have 2 questions, expecting your answers.

1. The statistc is for prewhere conditions optimization, have you considered expanding it to other scenarios?
2. As my understanding it can be used only in local execution node, am I right?

**Comment by hanfei1991** [comment]
@JackyWoo 
1. yes, statistics can be used in any where if we define the interface properly.
2. It works for local execution but also can be used in any other ways as long as we can collect statistics for every part we need
3. It only supports integer for now because I want to keep this PR simple. Let's do it step by step.

**Comment by JackyWoo** [comment]
thanks

**Review by CurtizJ — APPROVED** [review]
LGTM as a first step. Let's continue working on tasks from the [backlog](https://github.com/ClickHouse/ClickHouse/issues/55065).

**Comment by hanfei1991** [comment]
`00002_log_and_exception_messages_formatting` will be okay if we run normal stateless tests

**Comment by JackyWoo** [comment]
@hanfei1991 how to support multi-types of statistics such as hyperloglog, cm-sketch in SQL? 
1. Compatible with the current version
```
CREATE TABLE test (
    a Int64 STATISTIC type (tdigest, hyperloglog)
) ...

ALTER TABLE test ADD / DROP / CLEAR / MATERIALIZE STATISTIC a, b TYPE tdigest, hyperloglog
```

2. more flexible
```
CREATE TABLE test (
    a Int64 STATISTIC type (tdigest, hyperloglog)
) ...

ALTER TABLE test ADD / DROP / CLEAR / MATERIALIZE STATISTIC a TYPE tdigest, hyperloglog, b TYPE tdigest
```

which do you prefer or you have a better one?

**Comment by hanfei1991** [comment]
the 1st one is a amendment for the 3rd one @JackyWoo

**Pull request merged**